<?xml version="1.0" encoding="ISO-8859-15"?>
<launch>
  <!-- Input topic names for stereo dense depth reconstruction -->
  <arg name="left_cam_info_topic"  default="/tesse/left_cam/camera_info"/>
  <arg name="right_cam_info_topic" default="/tesse/right_cam/camera_info"/>
  <arg name="left_cam_topic"       default="/tesse/left_cam/image_raw"/>
  <arg name="right_cam_topic"      default="/tesse/right_cam/image_raw"/>

  <!-- StereoBM parameters -->
  <!-- Default params are tuned for left cam depth image in TESSE -->
  <arg name="texture_threshold"        default="1"/>
  <arg name="disparity_range"          default="70"/>
  <arg name="min_disparity"            default="-35"/>
  <arg name="correlation_window_size"  default="21"/>

  <!-- Run stereo_dense_reconstruction_node -->
  <arg name="run_stereo_dense"     default="true"/>
  <group ns="stereo_gray" if="$(arg run_stereo_dense)">
    <!-- StereoBM -->
    <node name="stereo_image_proc" pkg="stereo_image_proc"
      type="stereo_image_proc" clear_params="true" output="screen"
      if="$(arg run_stereo_dense)">
      <remap from="left/image_rect"   to="/dev/null/left_cam"/>
      <remap from="right/image_rect"  to="/dev/null/right_cam"/>
      <!-- Input -->
      <remap from="left/image_raw"          to="$(arg left_cam_topic)"/>
      <remap from="left/camera_info"        to="$(arg left_cam_info_topic)"/>
      <remap from="right/image_raw"         to="$(arg right_cam_topic)"/>
      <remap from="right/camera_info"       to="$(arg right_cam_info_topic)"/>
      <!-- Output -->
      <remap from="/stereo_gray/points2"    to="/stereo_gray/points2"/>
      <remap from="/stereo_gray/disparity"  to="/stereo_gray/disparity"/>
      <!-- Params -->
      <param name="approximate_sync"         value="true" />
      <param name="texture_threshold"        value="$(arg texture_threshold)"/>
      <param name="disparity_range"          value="$(arg disparity_range)"/>
      <param name="min_disparity"            value="$(arg min_disparity)"/>
      <param name="correlation_window_size"  value="$(arg correlation_window_size"/>
    </node>
    <!-- Run converter from the stereo disparity image to a depth image -->
    <arg name="depth_img"      value="depth/image_raw"/>
    <node pkg="nodelet" type="nodelet" name="stereo_proc"  args="manager"/>
    <node name="disparity_image_proc" pkg="nodelet" type="nodelet"
      args="load disparity_image_proc/depth_image stereo_proc" output="screen">
      <!-- Inputs -->
      <remap from="disparity"             to="/stereo_gray/disparity"/>
      <remap from="left/camera_info"      to="$(arg left_cam_info_topic)"/>
      <remap from="right/camera_info"     to="$(arg right_cam_info_topic)"/>
      <!-- Output -->
      <remap from="~depth_image"          to="~$(arg depth_img)"/>
      <!-- Params -->
      <param name="queue_size"         value="5"/>
      <param name="approximate_sync"   value="true"/>
    </node>
  </group>
</launch>
